tracer/src/Datadog.Trace/Debugger/Expressions/ProbeExpressionParser.cs

```diff
@@ -0,0 +1,483 @@
+        }
+
+        if (typeof(T) != typeof(string))
+        {
+            // let the caller throw the correct assign exception
+            return finalExpr;
+        }
+
+        // for string, call ToString or return the type name
+        if (SupportedTypesService.IsSafeToCallToString(finalExpr.Type))
+        {
+            finalExpr = Expression.Call(finalExpr, GetMethodByReflection(typeof(object), nameof(object.ToString), Type.EmptyTypes));
+        }
+        else
+        {
+            finalExpr = Expression.Constant(finalExpr.Type.FullName, typeof(string));
+        }
+
+        return finalExpr;
+    }
+
+    private ExpressionBodyAndParameters ParseProbeExpression(string expressionJson, ScopeMember @this, ScopeMember[] argsOrLocals)
+    {
+        @this.Type ??= @this.Value?.GetType();
+        if (string.IsNullOrEmpty(expressionJson) || argsOrLocals == null || @this.Type == null)
+        {
+            throw new ArgumentException($"{nameof(ParseProbeExpression)} has been called with an invalid argument");
+        }
+
+        var scopeMembers = new List<ParameterExpression>();
+        var expressions = new List<Expression>();
+        var thisParameterExpression = Expression.Parameter(@this.GetType());
+        var thisVariable = Expression.Variable(@this.Type, "this");
+        expressions.Add(Expression.Assign(thisVariable, Expression.Convert(Expression.Field(thisParameterExpression, "Value"), @this.Type)));
+        scopeMembers.Add(thisVariable);
+
+        var argsOrLocalsParameterExpression = Expression.Parameter(argsOrLocals.GetType());
+
+        for (var index = 0; index < argsOrLocals.Length; index++)
+        {
+            if (argsOrLocals[index].Type == null)
+            {
+                break;
+            }
+
+            var argOrLocal = argsOrLocals[index];
+            var variable = Expression.Variable(argOrLocal.Type, argOrLocal.Name);
+            scopeMembers.Add(variable);
+
+            expressions.Add(
+                Expression.Assign(
+                    variable,
+                    Expression.Convert(
+                    Expression.Field(
+                    Expression.ArrayIndex(
+                        argsOrLocalsParameterExpression,
+                        Expression.Constant(index)),
+                    "Value"),
+                    argOrLocal.Type)));
+        }
+
+        var reader = new JsonTextReader(new StringReader(expressionJson));
+        SetReaderAtExpressionStart(reader);
+
+        var result = Expression.Variable(typeof(T), "$dd_el_result");
+        scopeMembers.Add(result);
+        var finalExpr = ParseRoot(reader, scopeMembers);
+        finalExpr = HandleReturnType(finalExpr);
+        expressions.Add(finalExpr is not GotoExpression ? Expression.Assign(result, finalExpr) : finalExpr);
+        expressions.Add(Expression.Label(ReturnTarget, result));
+        var body = (Expression)Expression.Block(scopeMembers, expressions);
+        if (body.CanReduce)
```

Comment:
You might want to consider `while` this expression (with certain `tryCount` for safeguard from infinite loop) as `ReduceAndCheck` might return a node which can/must be reduced.